### PR TITLE
Feature/cdap 14770 migrate lineage dataset

### PR DESCRIPTION
**Pull request by jayfjin**
Migrate LineageDataset to the new SPI.

Also ensure system tables in the new SPI are correctly handled by the class loader. Prevent the classloader from writing lineage to system tables.

Build: 
https://builds.cask.co/browse/CDAP-DUT6817-5

### CODE REVIEW

**Comment by jayfjin** [comment]
Build is green https://builds.cask.co/browse/CDAP-DUT6817-5

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java [hunk 1]
```diff
@@ -91,21 +95,20 @@
   @After
   public void cleanup() throws Exception {
     deleteNamespace(NamespaceId.DEFAULT.getNamespace());
+    TransactionRunner transactionRunner = getInjector().getInstance(TransactionRunner.class);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java:97]: why is this required? It looks like it was previously not done. Do the test cases not leave a clean state behind?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java [hunk 2]
```diff
@@ -91,21 +95,20 @@
...
-                                                                     getTxClient(), lineageDatasetId);
-    LineageWriter lineageWriter = new BasicLineageWriter(getDatasetFramework(), getTxClient()) {
-      @Override
-      protected DatasetId getLineageDatasetId() {
-        return lineageDatasetId;
-      }
-    };
+    TransactionRunner transactionRunner = getInjector().getInstance(TransactionRunner.class);
+    LineageStoreReader lineageReader =
+      new DefaultLineageStoreReader(transactionRunner);
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java]: unnecessary line break?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java [hunk 3]
```diff
@@ -208,15 +211,10 @@ public void testSimpleLoopLineage() throws Exception {
...
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java]: same here

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java [hunk 4]
```diff
@@ -280,15 +278,10 @@ public void testDirectCycle() throws Exception {
...
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java]: ditto

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java (hunk 1 above)
Comment (anew) [reply cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java:97]: oh is it because previously each test case used a different name for the lineage dataset. So they never cleaned up after themselves.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java [hunk 5]
```diff
@@ -170,9 +170,8 @@ public static DatasetSpecification fixOriginalProperties(@Nullable DatasetSpecif
         props.remove(Table.PROPERTY_SCHEMA);
         props.remove(Table.PROPERTY_SCHEMA_ROW_FIELD);
 
-        // LineageDataset and UsageDataset add the conflict level of none
-      } else if (UsageDataset.class.getSimpleName().equals(type) ||
-        LineageDataset.class.getName().equals(type) || "lineageDataset".equals(type)) {
+        // UsageDataset add the conflict level of none
+      } else if (UsageDataset.class.getSimpleName().equals(type)) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java:174]: if I understand correctly, the old dataset had conflict detection turned off. How do we deal with conflicts now?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+  private StructuredTable programTable;
+
+  /**
+   * Gets an instance of {@link LineageTable}.
+   *
+   * @param context the {@link StructuredTableContext} for getting the dataset instance.
+   * @return an instance of {@link LineageTable}
+   */
+  @VisibleForTesting
+  public static LineageTable getLineageDataset(StructuredTableContext context) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: should this be called getLineageTable()?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+  private static final Logger LOG = LoggerFactory.getLogger(LineageTable.class);
+
+  private final StructuredTableContext structuredTableContext;
+  private StructuredTable datasetTable;
+  private StructuredTable programTable;
+
+  /**
+   * Gets an instance of {@link LineageTable}.
+   *
+   * @param context the {@link StructuredTableContext} for getting the dataset instance.
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: should just say, for getting the lineage table (we don't know whether it is a dataset)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 6 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: or, I guess in other tables, we called this create()

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+  /**
+   * @return a set of entities (program and data it accesses) associated with a program run.
+   */
+  public Set<NamespacedEntityId> getEntitiesForRun(ProgramRunId run) throws IOException {
+    ImmutableSet.Builder<NamespacedEntityId> recordBuilder = ImmutableSet.builder();
+    List<Field<?>> prefix = getRunScanStartKey(run);
+    try (CloseableIterator<StructuredRow> iterator = getProgramTable().scan(Range.singleton(prefix), Integer.MAX_VALUE)) {
+      while (iterator.hasNext()) {
+        StructuredRow row = iterator.next();
+        if (run.getEntityName().equals(row.getString(StoreDefinition.LineageStore.RUN_FIELD))) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: Nit: I prefer run.getRun() over the generic getEntityName()

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    List<Field<?>> programFields = getProgramKey(run, datasetInstance, accessType);
+    addAccessTime(programFields, accessTimeMillis);
+    getProgramTable().upsert(programFields);
+  }
+
+  /**
+   * @return a set of entities (program and data it accesses) associated with a program run.
+   */
+  public Set<NamespacedEntityId> getEntitiesForRun(ProgramRunId run) throws IOException {
+    ImmutableSet.Builder<NamespacedEntityId> recordBuilder = ImmutableSet.builder();
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: I would call it `setBuilder` or just `builder` - it does not build a record.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+   *
+   * @param program program for which to fetch access information
+   * @param start start time period
+   * @param end end time period
+   * @param filter filter to be applied on result set
+   * @return program-dataset access information
+   */
+  public Set<Relation> getRelations(ProgramId program, long start, long end, Predicate<Relation> filter)
+    throws IOException {
+    return scanRelations(getProgramTable(),
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java:170]: smart! One method to scan either table. 💯

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                         getProgramScanEndKey(program, start),
+                         filter);
+  }
+
+  /**
+   * @return a set of access times (for program and data it accesses) associated with a program run.
+   */
+  @VisibleForTesting
+  public List<Long> getAccessTimesForRun(ProgramRunId run) throws IOException {
+    ImmutableList.Builder<Long> recordBuilder = ImmutableList.builder();
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: same here, it does not build a record. I assume this is from copy&pasting :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+    List<Field<?>> prefix = getRunScanStartKey(run);
+    try (
+      CloseableIterator<StructuredRow> iterator = getProgramTable().scan(Range.singleton(prefix), Integer.MAX_VALUE)) {
+      while (iterator.hasNext()) {
+        StructuredRow row = iterator.next();
+        if (run.getEntityName().equals(row.getString(StoreDefinition.LineageStore.RUN_FIELD))) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: run.getRun()?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+  private void addAccessTime(List<Field<?>> fields, long accessTime) {
+    fields.add(Fields.longField(StoreDefinition.LineageStore.ACCESS_TIME_FIELD, accessTime));
+  }
+
+  private List<Field<?>> getDatasetScanKey(DatasetId datasetInstance, long time) {
+    long invertedStartTime = invertTime(time);
+    List<Field<?>> fields = new ArrayList<>();
+    addDataset(fields, datasetInstance);
+    fields.add(Fields.longField(StoreDefinition.LineageStore.START_TIME_FIELD, invertedStartTime));
+
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: extra newline?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 14]
```diff
@@ -0,0 +1,343 @@
...
+  }
+
+  private void addDataset(List<Field<?>> fields, DatasetId datasetInstance) {
+      fields.add(Fields.stringField(StoreDefinition.LineageStore.NAMESPACE_FIELD, datasetInstance.getNamespace()));
+      fields.add(Fields.stringField(StoreDefinition.LineageStore.DATASET_FIELD, datasetInstance.getEntityName()));
+  }
+
+  private void addProgram(List<Field<?>> fields, ProgramId program) {
+      fields.add(Fields.stringField(StoreDefinition.LineageStore.PROGRAM_NAMESPACE_FIELD, program.getNamespace()));
+      fields.add(Fields.stringField(StoreDefinition.LineageStore.PROGRAM_PARENT_FIELD,
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: PROGRAM_PARENT? Isn't that the application? 
Also, applications can have multiple versions. Does lineage not record the version?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+    return new DatasetId(row.getString(StoreDefinition.LineageStore.NAMESPACE_FIELD),
+                         row.getString(StoreDefinition.LineageStore.DATASET_FIELD));
+  }
+
+  private long invertTime(long time) {
+    return Long.MAX_VALUE - time;
+  }
+
+  private long getInvertedStartTime(ProgramRunId run) {
+    return invertTime(RunIds.getTime(RunIds.fromString(run.getEntityName()), TimeUnit.MILLISECONDS));
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: run.getRun()?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+      public void emitAudit(AccessType accessType) {
+        // no-op
+      }
+    };
+  private final DatasetDefinition datasetDefinition;
+
+  public NoSqlStructuredTableDatasetDefinition(DatasetDefinition datasetDefinition) {
+    this.datasetDefinition = datasetDefinition;
+  }
+
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java:58]: extra newline?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+  @Override
+  public Dataset getDataset(DatasetContext datasetContext, DatasetSpecification spec, Map arguments,
+                            ClassLoader classLoader) throws IOException {
+    try {
+      return DefaultDatasetRuntimeContext.execute(SYSTEM_NAMESPACE_ENFORCER, SYSTEM_NAMESPACE_ACCESS_RECORDER,
+                                                  null, null, null,
+                                                  () -> datasetDefinition.getDataset(
+                                                    datasetContext, spec, arguments, classLoader));
+    } catch (IOException e) {
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java]: I think this can be written as:
```
catch (Exception e) {
  Throwables.propagateIfPossible(e, IOException.class);
  throw Throwables.propagate(e);
}
```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+  public DatasetAdmin getAdmin(DatasetContext datasetContext, DatasetSpecification spec, ClassLoader classLoader)
+    throws IOException {
+    return datasetDefinition.getAdmin(datasetContext, spec, classLoader);
+  }
+
+  @Override
+  public Dataset getDataset(DatasetContext datasetContext, DatasetSpecification spec, Map arguments,
+                            ClassLoader classLoader) throws IOException {
+    try {
+      return DefaultDatasetRuntimeContext.execute(SYSTEM_NAMESPACE_ENFORCER, SYSTEM_NAMESPACE_ACCESS_RECORDER,
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java:79]: I am not exactly sure how this works. Can you ask @chtyim to review this part? thanks

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 19]
```diff
@@ -601,4 +605,62 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
...
+   * Schema for {@link LineageTable}.
+   */
+  public static final class LineageStore {
+    public static final StructuredTableId DATASET_LINEAGE_TABLE = new StructuredTableId("dataset_lineage");
+    public static final StructuredTableId PROGRAM_LINEAGE_TABLE = new StructuredTableId("program_lineage");
+    public static final String NAMESPACE_FIELD = "namespace";
+    public static final String DATASET_FIELD = "dataset";
+    public static final String START_TIME_FIELD = "start_time";
+    public static final String PROGRAM_NAMESPACE_FIELD = "program_namespace";
+    public static final String PROGRAM_PARENT_FIELD = "program_parent";
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: this should be PROGRAM_APPLICATION_FIELD and program_app

Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: also (see my comment above) where is the program version recorded?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 20]
```diff
@@ -601,4 +605,62 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
...
+    public static final String PROGRAM_TYPE_FIELD = "program_type";
+    public static final String PROGRAM_FIELD = "program";
+    public static final String RUN_FIELD = "run";
+    public static final String ACCESS_TYPE_FIELD = "access_type";
+    public static final String ACCESS_TIME_FIELD = "access_time";
+
+
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: bunch of extra newlines?

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java [hunk 21]
```diff
@@ -601,4 +605,62 @@ public static void createTables(StructuredTableAdmin tableAdmin) throws IOExcept
...
+
+    public static final StructuredTableSpecification DATASET_LINEAGE_SPEC = new StructuredTableSpecification.Builder()
+      .withId(DATASET_LINEAGE_TABLE)
+      .withFields(Fields.stringType(NAMESPACE_FIELD),
+                  Fields.stringType(DATASET_FIELD),
+                  Fields.longType(START_TIME_FIELD),
+                  Fields.stringType(PROGRAM_NAMESPACE_FIELD),
+                  Fields.stringType(PROGRAM_PARENT_FIELD),
```
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: similar here, it's the application not the parent

File: cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/lineage/SqlLineageTableTest.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+import java.io.IOException;
+import javax.sql.DataSource;
+
+public class SqlLineageTableTest extends LineageTableTest {
+
+  private static EmbeddedPostgres pg;
+
+  @BeforeClass
+  public static void beforeClass() throws IOException, TableAlreadyExistsException {
+    pg = EmbeddedPostgres.start();
```
Comment (anew) [inline cdap-data-fabric/src/test/java/co/cask/cdap/data2/metadata/lineage/SqlLineageTableTest.java]: please see here for how to set the temp directory correctly for the embedded postgres: https://github.com/cdapio/cdap/pull/11063/files#diff-144810ba58266142b9892beffc9c38c6R57

**Review by anew — CHANGES_REQUESTED** [review]

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java (hunk 1 above)
Comment (jayfjin) [reply cdap-app-fabric/src/test/java/co/cask/cdap/metadata/LineageAdminTest.java:97]: yes correct

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java (hunk 5 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java:174]: We do not. From our discussion in person - we do not expect conflicts anymore - and verified through usages of LineageTable.addAccess

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 11 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: This was a class previously so a lot of the names I didn't change from the previous implementation :)

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 14 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: I did not change what fields the lineage did and did not record - it's the same as the previous implementation

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java (hunk 17 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java]: thanks

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java (hunk 18 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java:79]: sg

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 19 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: unless I am mistaken it was never recorded?

**Comment by jayfjin** [comment]
Addressed comments, will also ping Terence for review on the classloader commit

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java (hunk 18 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java:79]: Terence pinged, "lgtm for that commit"

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java (hunk 5 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetsUtil.java:174]: got it

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java (hunk 18 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTableDatasetDefinition.java:79]: 👍

File: cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java (hunk 19 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/store/StoreDefinition.java]: oh bummer. So this was never there. Or maybe on purpose? In that case, ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 11 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: I see. I did not compare this to the previous impl, this file showed up as a new file in the diff.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 14 above)
Comment (anew) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: the previous implementation? I guess that called `program.getParent().getEntityName()`, but it did not name the fields in the row key. Here we give names to the columns of the primary key, so we should name them accurately.

**Review by anew — APPROVED** [review]
LGTM. Nice job 👍

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java (hunk 14 above)
Comment (jayfjin) [reply cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageTable.java]: ah yes, was referring to the comment about lineage not recording the version, not the name :p
